File: internal/sdk/typescript/src/approvals.ts
import { createHash, timingSafeEqual } from "node:crypto";
import { clone, create, type MessageInitShape, toBinary } from "@bufbuild/protobuf";

import type { ResourceRef } from "../../../../protocols/generated/typescript/common/v1/resource_reference_pb.js";
import {
	ConsumeApprovalRequestSchema,
	DecideApprovalRequestSchema,
	GetApprovalRequestRequestSchema,
	ListApprovalRequestsRequestSchema,
	type ListApprovalRequestsResponse,
	RequestApprovalRequestSchema,
} from "../../../../protocols/generated/typescript/internal/workflow/v1/workflow_service_pb.js";
import {
	ApprovalBindingSchema,
	ApprovalDecisionValue,
	type ApprovalReceipt,
	ApprovalReceiptSchema,
	type ApprovalRequest,
	ApprovalRequestSchema,
} from "../../../../protocols/generated/typescript/workflow/v1/approval_pb.js";
import type { ClientCore } from "./core.js";
import { MindcladeError } from "./error.js";
import { listPage, type Page, withPageToken } from "./pagination.js";
import {
	commandContext,
	type ListOptions,
	prepareCall,
	type SdkCallOptions,
	type SubmitOptions,
} from "./request.js";
import { invokeUnary } from "./retry.js";

const REQUEST = "/mindclade.internal.workflow.v1.ApprovalService/RequestApproval";
const GET = "/mindclade.internal.workflow.v1.ApprovalService/GetApprovalRequest";
const LIST = "/mindclade.internal.workflow.v1.ApprovalService/ListApprovalRequests";
const DECIDE = "/mindclade.internal.workflow.v1.ApprovalService/DecideApproval";
const CONSUME = "/mindclade.internal.workflow.v1.ApprovalService/ConsumeApproval";
const SHA256 = /^sha256:[0-9a-f]{64}$/;
const RESOURCE_ID = /^[A-Za-z0-9._-]{1,128}$/;
const MAXIMUM_PAGE_SIZE = 200;

/** Exact-intent approval and immutable receipt facade over generated contracts. */
export class Approvals {
	readonly #core: ClientCore;

	constructor(core: ClientCore) {
		this.#core = core;
	}

	async request(
		input: MessageInitShape<typeof ApprovalRequestSchema>,
		options: SubmitOptions,
	): Promise<ApprovalRequest> {
		ensureUnfenced(options);
		const approval = create(ApprovalRequestSchema, input);
		if (approval.binding === undefined) {
			throw MindcladeError.invalidArgument("approval request requires a generated binding");
		}
		if (approval.name !== "") scopedName(this.#core, approval.name, "approvalRequests");
		normalizeScope(this.#core, approval);
		approval.requestedByPrincipalRef = this.#core.config.identity.principalId;
		if (approval.binding.tool !== undefined) normalizeReference(this.#core, approval.binding.tool);
		for (const decision of approval.policyDecisions) {
			if (decision.resource !== undefined) normalizeReference(this.#core, decision.resource);
		}
		verifyBinding(approval.binding);
		delete approval.context;
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		approval.context = {
			...commandContext(this.#core.config, prepared, options),
			canonicalRequestDigest: sha256(toBinary(ApprovalRequestSchema, approval)),
		};
		const expectedBinding = approval.binding.bindingDigest;
		const response = await invokeUnary(
			this.#core,
			prepared,
			REQUEST,
			options.idempotencyKey,
			(call) =>
				this.#core.raw.approvals.requestApproval(
					create(RequestApprovalRequestSchema, { approvalRequest: approval }),
					call,
				),
		);
		if (response.approvalRequest === undefined) {
			throw MindcladeError.protocol("RequestApproval response omitted its request");
		}
		const created = clone(ApprovalRequestSchema, response.approvalRequest);
		scopedName(this.#core, created.name, "approvalRequests");
		if (
			created.binding === undefined ||
			!safeEqual(created.binding.bindingDigest, expectedBinding)
		) {
			throw MindcladeError.protocol("RequestApproval returned inconsistent durable intent");
		}
		return created;
	}

	async get(name: string, options: SdkCallOptions = {}): Promise<ApprovalRequest> {
		ensureUnfenced(options);
		const request = create(GetApprovalRequestRequestSchema, {
			name: scopedName(this.#core, name, "approvalRequests"),
		});
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			GET,
			undefined,
			(call) => this.#core.raw.approvals.getApprovalRequest(request, call),
		);
		if (response.approvalRequest === undefined) {
			throw MindcladeError.protocol("GetApprovalRequest response omitted its request");
		}
		return clone(ApprovalRequestSchema, response.approvalRequest);
	}

	/** Returns the first page, which also iterates the whole cursor. */
	async list(
		input: MessageInitShape<typeof ListApprovalRequestsRequestSchema> = {},
		options: ListOptions = {},
	): Promise<Page<ApprovalRequest, ListApprovalRequestsResponse>> {
		ensureUnfenced(options);
		const request = create(ListApprovalRequestsRequestSchema, input);
		request.parent = normalizeParent(this.#core, request.parent);
		if (
			request.page !== undefined &&
			(!Number.isInteger(request.page.pageSize) ||
				request.page.pageSize < 0 ||
				request.page.pageSize > MAXIMUM_PAGE_SIZE)
		) {
			throw MindcladeError.invalidArgument(
				"approval page size must be an integer between zero and 200",
			);
		}
		return await listPage({
			cursor: (response) => response.page?.nextPageToken ?? "",
			fetch: async (pageToken) => {
				const paged = withPageToken(ListApprovalRequestsRequestSchema, request, pageToken);
				const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
				const response = await invokeUnary(
					this.#core,
					prepared,
					LIST,
					undefined,
					(call) => this.#core.raw.approvals.listApprovalRequests(paged, call),
				);
				return { requestId: prepared.requestId, response };
			},
			items: (response) => response.approvalRequests,
			limits: options.limits,
			pageSize: request.page?.pageSize ?? 0,
			pageToken: request.page?.pageToken ?? "",
			signal: options.signal,
		});
	}

	async decide(
		input: MessageInitShape<typeof DecideApprovalRequestSchema>,
		options: SubmitOptions,
	): Promise<ApprovalReceipt> {
		ensureUnfenced(options);
		const request = create(DecideApprovalRequestSchema, input);
		request.name = scopedName(this.#core, request.name, "approvalRequests");
		if (
			request.etag.trim() === "" ||
			request.decision === ApprovalDecisionValue.UNSPECIFIED ||
			request.reasonCode.trim() === "" ||
			request.safeReason.length > 2048
		) {
			throw MindcladeError.invalidArgument(
				"approval decision requires an ETag, decision, reason code, and bounded safe reason",
			);
		}
		delete request.context;
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		request.context = {
			...commandContext(this.#core.config, prepared, options),
			canonicalRequestDigest: sha256(toBinary(DecideApprovalRequestSchema, request)),
		};
		const response = await invokeUnary(
			this.#core,
			prepared,
			DECIDE,
			options.idempotencyKey,
			(call) => this.#core.raw.approvals.decideApproval(request, call),
		);
		return validateDecisionReceipt(this.#core, request, response.approvalReceipt);
	}

	async consume(
		input: MessageInitShape<typeof ConsumeApprovalRequestSchema>,
		options: SubmitOptions,
	): Promise<ApprovalReceipt> {
		ensureUnfenced(options);
		const request = create(ConsumeApprovalRequestSchema, input);
		request.receiptName = scopedName(this.#core, request.receiptName, "approvalReceipts");
		if (
			!SHA256.test(request.bindingDigest) ||
			request.callId.trim() === "" ||
			request.callId.length > 1024
		) {
			throw MindcladeError.invalidArgument(
				"approval consumption requires a binding digest and bounded call ID",
			);
		}
		delete request.context;
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		request.context = {
			...commandContext(this.#core.config, prepared, options),
			canonicalRequestDigest: sha256(toBinary(ConsumeApprovalRequestSchema, request)),
		};
		const response = await invokeUnary(
			this.#core,
			prepared,
			CONSUME,
			options.idempotencyKey,
			(call) => this.#core.raw.approvals.consumeApproval(request, call),
		);
		const receipt = response.approvalReceipt;
		if (
			receipt === undefined ||
			receipt.name !== request.receiptName ||
			receipt.consumedAt === undefined ||
			receipt.consumedByCallId !== request.callId ||
			receipt.binding === undefined ||
			!safeEqual(receipt.binding.bindingDigest, request.bindingDigest) ||
			!SHA256.test(receipt.receiptDigest)
		) {
			throw MindcladeError.protocol("ConsumeApproval returned an inconsistent receipt");
		}
		return clone(ApprovalReceiptSchema, receipt);
	}
}

const verifyBinding = (binding: ReturnType<typeof create<typeof ApprovalBindingSchema>>): void => {
	if (
		!SHA256.test(binding.intentDigest) ||
		!SHA256.test(binding.parametersDigest) ||
		!SHA256.test(binding.bindingDigest)
	) {
		throw MindcladeError.invalidArgument("approval binding requires canonical SHA-256 digests");
	}
	const unsigned = clone(ApprovalBindingSchema, binding);
	const supplied = unsigned.bindingDigest;
	unsigned.bindingDigest = "";
	if (!safeEqual(supplied, sha256(toBinary(ApprovalBindingSchema, unsigned)))) {
		throw MindcladeError.invalidArgument(
			"approval binding digest does not match its generated payload",
		);
	}
};

const validateDecisionReceipt = (
	core: ClientCore,
	request: ReturnType<typeof create<typeof DecideApprovalRequestSchema>>,
	receipt: ApprovalReceipt | undefined,
): ApprovalReceipt => {
	if (
		receipt === undefined ||
		receipt.request === undefined ||
		receipt.binding === undefined ||
		receipt.decidedAt === undefined ||
		receipt.request.name !== request.name ||
		receipt.decision !== request.decision ||
		receipt.reasonCode !== request.reasonCode ||
		receipt.safeReason !== request.safeReason ||
		!SHA256.test(receipt.receiptDigest)
	) {
		throw MindcladeError.protocol("DecideApproval returned an inconsistent receipt");
	}
	scopedName(core, receipt.name, "approvalReceipts");
	normalizeReference(core, receipt.request, "approval_request", "approvalRequests");
	return clone(ApprovalReceiptSchema, receipt);
};

const normalizeReference = (
	core: ClientCore,
	reference: ResourceRef,
	expectedType?: string,
	collection?: string,
): void => {
	const name =
		collection === undefined
			? sameProjectResource(core, reference.name)
			: scopedName(core, reference.name, collection);
	const resourceId = name.slice(name.lastIndexOf("/") + 1);
	if (reference.resourceId !== "" && reference.resourceId !== resourceId) {
		throw MindcladeError.invalidArgument("approval reference ID does not match its name");
	}
	if (expectedType !== undefined) {
		if (reference.resourceType !== "" && reference.resourceType !== expectedType) {
			throw MindcladeError.invalidArgument("approval reference type does not match its use");
		}
		reference.resourceType = expectedType;
	} else if (reference.resourceType.trim() === "") {
		throw MindcladeError.invalidArgument("approval reference type is required");
	}
	reference.resourceId = resourceId;
	normalizeScope(core, reference);
};

const normalizeScope = (core: ClientCore, value: { tenantId: string; projectId: string }): void => {
	if (
		(value.tenantId !== "" && value.tenantId !== core.config.identity.tenantId) ||
		(value.projectId !== "" && value.projectId !== core.config.identity.projectId)
	) {
		throw MindcladeError.invalidArgument("approval resource scope conflicts with client identity");
	}
	value.tenantId = core.config.identity.tenantId;
	value.projectId = core.config.identity.projectId;
};

const normalizeParent = (core: ClientCore, parent: string): string => {
	const expected = projectName(core);
	if (parent !== "" && parent !== expected) {
		throw MindcladeError.invalidArgument("approval parent does not match client scope");
	}
	return expected;
};

const projectName = (core: ClientCore): string => {
	const tenant = core.config.identity.tenantId.startsWith("tenants/")
		? core.config.identity.tenantId
		: `tenants/${core.config.identity.tenantId}`;
	const project = core.config.identity.projectId;
	if (project.startsWith("tenants/")) return project;
	return project.startsWith("projects/") ? `${tenant}/${project}` : `${tenant}/projects/${project}`;
};

const scopedName = (core: ClientCore, name: string, collection: string): string => {
	const prefix = `${projectName(core)}/${collection}/`;
	const id = name.startsWith(prefix) ? name.slice(prefix.length) : "";
	if (!RESOURCE_ID.test(id)) {
		throw MindcladeError.invalidArgument(`${collection} resource is outside client scope`);
	}
	return name;
};

const sameProjectResource = (core: ClientCore, name: string): string => {
	const prefix = `${projectName(core)}/`;
	const relative = name.startsWith(prefix) ? name.slice(prefix.length) : "";
	const parts = relative.split("/");
	if (parts.length !== 2 || !RESOURCE_ID.test(parts[1] ?? "")) {
		throw MindcladeError.invalidArgument("approval reference is outside client scope");
	}
	return name;
};

const ensureUnfenced = (options: SdkCallOptions): void => {
	if (options.leaseToken !== undefined || options.workerId !== undefined) {
		throw MindcladeError.invalidArgument(
			"worker and lease credentials are not accepted by approval RPCs",
		);
	}
};

const safeEqual = (left: string, right: string): boolean => {
	const first = Buffer.from(left, "utf8");
	const second = Buffer.from(right, "utf8");
	return first.length === second.length && timingSafeEqual(first, second);
};

const sha256 = (value: Uint8Array): string =>
	`sha256:${createHash("sha256").update(value).digest("hex")}`;

File: internal/sdk/rust/src/approvals.rs
use std::sync::Arc;

use mindclade_protocols::{
    internal::workflow::v1::{
        ConsumeApprovalRequest, DecideApprovalRequest, GetApprovalRequestRequest,
        ListApprovalRequestsRequest, RequestApprovalRequest,
    },
    workflow::v1::{ApprovalBinding, ApprovalDecisionValue, ApprovalReceipt, ApprovalRequest},
};
use prost::Message;
use sha2::{Digest, Sha256};

use crate::{
    CallOptions, ClientCore, Error, Page, Pages, SubmitOptions,
    request::{initial_page_token, page_request},
    retry::registered_method_safety,
    workflows::{command_context, normalize_parent, valid_sha256, validate_page, workflow_name},
};

const REQUEST: &str = "/mindclade.internal.workflow.v1.ApprovalService/RequestApproval";
const GET: &str = "/mindclade.internal.workflow.v1.ApprovalService/GetApprovalRequest";
const LIST: &str = "/mindclade.internal.workflow.v1.ApprovalService/ListApprovalRequests";
const DECIDE: &str = "/mindclade.internal.workflow.v1.ApprovalService/DecideApproval";
const CONSUME: &str = "/mindclade.internal.workflow.v1.ApprovalService/ConsumeApproval";

/// Exact-intent approval and immutable receipt API over generated contracts.
#[derive(Clone)]
pub struct Approvals {
    core: Arc<ClientCore>,
}

impl Approvals {
    pub(crate) fn new(core: Arc<ClientCore>) -> Self {
        Self { core }
    }

    /// Records exact generated approval intent after validating its canonical
    /// binding digest and replacing caller-supplied identity/context.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid binding/scope/context or RPC response.
    pub async fn request(
        &self,
        mut approval: ApprovalRequest,
        options: SubmitOptions,
    ) -> Result<ApprovalRequest, Error> {
        let binding = approval
            .binding
            .as_ref()
            .ok_or_else(|| Error::invalid_argument("approval binding is required"))?;
        verify_binding(binding)?;
        if (!approval.tenant_id.is_empty()
            && approval.tenant_id != self.core.config.identity.tenant_id())
            || (!approval.project_id.is_empty()
                && approval.project_id != self.core.config.identity.project_id())
        {
            return Err(Error::invalid_argument(
                "approval request scope does not match client",
            ));
        }
        approval.tenant_id = self.core.config.identity.tenant_id().to_owned();
        approval.project_id = self.core.config.identity.project_id().to_owned();
        approval.requested_by_principal_ref = self.core.config.identity.principal_id().to_owned();
        approval.context = None;
        let prepared = options.call.prepare(&self.core.config);
        approval.context = Some(command_context(&self.core, &prepared, &options, &approval)?);
        let expected_digest = approval
            .binding
            .as_ref()
            .map(|value| value.binding_digest.clone())
            .unwrap_or_default();
        let key = options.idempotency_key.clone();
        let created = self
            .core
            .unary(
                RequestApprovalRequest {
                    approval_request: Some(approval),
                },
                &prepared,
                registered_method_safety(REQUEST),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.request_approval(request).await })
                },
            )
            .await?
            .into_inner()
            .approval_request
            .ok_or_else(|| Error::protocol("RequestApproval response omitted its request"))?;
        workflow_name(&self.core, &created.name, "approvalRequests")?;
        if created
            .binding
            .as_ref()
            .map(|value| value.binding_digest.as_str())
            != Some(expected_digest.as_str())
        {
            return Err(Error::protocol(
                "RequestApproval returned inconsistent durable intent",
            ));
        }
        Ok(created)
    }

    /// Reads one generated durable approval request.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope, transport, or response.
    pub async fn get(
        &self,
        name: impl Into<String>,
        options: CallOptions,
    ) -> Result<ApprovalRequest, Error> {
        let name = workflow_name(&self.core, &name.into(), "approvalRequests")?;
        let prepared = options.prepare(&self.core.config);
        self.core
            .unary(
                GetApprovalRequestRequest { name },
                &prepared,
                registered_method_safety(GET),
                None,
                |transport, request| {
                    Box::pin(async move { transport.get_approval_request(request).await })
                },
            )
            .await?
            .into_inner()
            .approval_request
            .ok_or_else(|| Error::protocol("GetApprovalRequest response omitted its request"))
    }

    /// Lists one bounded generated approval page while preserving its opaque
    /// server token.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope/pagination or RPC failure.
    pub fn list(
        &self,
        mut request: ListApprovalRequestsRequest,
        options: CallOptions,
    ) -> Result<Pages<ApprovalRequest>, Error> {
        normalize_parent(&self.core, &mut request.parent)?;
        validate_page(request.page.as_ref())?;
        let core = Arc::clone(&self.core);
        let token = initial_page_token(request.page.as_ref());
        Ok(Pages::new(
            move |page_token| {
                let core = Arc::clone(&core);
                let options = options.clone();
                let mut request = request.clone();
                async move {
                    request.page = Some(page_request(request.page.as_ref(), page_token));
                    let prepared = options.prepare(&core.config);
                    let response = core
                        .unary(
                            request,
                            &prepared,
                            registered_method_safety(LIST),
                            None,
                            |transport, request| {
                                Box::pin(
                                    async move { transport.list_approval_requests(request).await },
                                )
                            },
                        )
                        .await?;
                    let request_id = response.request_id().map(str::to_owned);
                    let response = response.into_inner();
                    Ok(Page::new(
                        response.approval_requests,
                        response.page,
                        response.read_time,
                        request_id,
                    ))
                }
            },
            token,
        ))
    }

    /// Records an independently authenticated decision under optimistic
    /// concurrency and validates the immutable returned receipt.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope/decision/ETag or inconsistent receipt.
    pub async fn decide(
        &self,
        mut request: DecideApprovalRequest,
        options: SubmitOptions,
    ) -> Result<ApprovalReceipt, Error> {
        request.name = workflow_name(&self.core, &request.name, "approvalRequests")?;
        if request.etag.trim().is_empty()
            || request.decision == ApprovalDecisionValue::Unspecified as i32
            || request.reason_code.trim().is_empty()
            || request.safe_reason.len() > 2048
        {
            return Err(Error::invalid_argument(
                "approval decision requires an ETag, decision, reason code, and bounded safe reason",
            ));
        }
        request.context = None;
        let prepared = options.call.prepare(&self.core.config);
        request.context = Some(command_context(&self.core, &prepared, &options, &request)?);
        let expected_name = request.name.clone();
        let expected_decision = request.decision;
        let expected_reason = request.reason_code.clone();
        let expected_safe_reason = request.safe_reason.clone();
        let key = options.idempotency_key.clone();
        let receipt = self
            .core
            .unary(
                request,
                &prepared,
                registered_method_safety(DECIDE),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.decide_approval(request).await })
                },
            )
            .await?
            .into_inner()
            .approval_receipt
            .ok_or_else(|| Error::protocol("DecideApproval response omitted its receipt"))?;
        workflow_name(&self.core, &receipt.name, "approvalReceipts")?;
        let request_ref = receipt
            .request
            .as_ref()
            .ok_or_else(|| Error::protocol("approval receipt omitted its request reference"))?;
        if request_ref.name != expected_name
            || receipt.decision != expected_decision
            || receipt.reason_code != expected_reason
            || receipt.safe_reason != expected_safe_reason
            || receipt.binding.is_none()
            || !valid_sha256(&receipt.receipt_digest)
            || !valid_timestamp(receipt.decided_at.as_ref())
        {
            return Err(Error::protocol(
                "DecideApproval returned an inconsistent receipt",
            ));
        }
        Ok(receipt)
    }

    /// Atomically consumes a generated receipt for one digest-bound call.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid binding/call identity or inconsistent receipt.
    pub async fn consume(
        &self,
        mut request: ConsumeApprovalRequest,
        options: SubmitOptions,
    ) -> Result<ApprovalReceipt, Error> {
        request.receipt_name =
            workflow_name(&self.core, &request.receipt_name, "approvalReceipts")?;
        if !valid_sha256(&request.binding_digest)
            || request.call_id.trim().is_empty()
            || request.call_id.len() > 512
        {
            return Err(Error::invalid_argument(
                "approval consumption requires a binding digest and bounded call ID",
            ));
        }
        request.context = None;
        let prepared = options.call.prepare(&self.core.config);
        request.context = Some(command_context(&self.core, &prepared, &options, &request)?);
        let expected_name = request.receipt_name.clone();
        let expected_digest = request.binding_digest.clone();
        let expected_call = request.call_id.clone();
        let key = options.idempotency_key.clone();
        let receipt = self
            .core
            .unary(
                request,
                &prepared,
                registered_method_safety(CONSUME),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.consume_approval(request).await })
                },
            )
            .await?
            .into_inner()
            .approval_receipt
            .ok_or_else(|| Error::protocol("ConsumeApproval response omitted its receipt"))?;
        if receipt.name != expected_name
            || receipt.consumed_by_call_id != expected_call
            || !valid_timestamp(receipt.consumed_at.as_ref())
            || receipt
                .binding
                .as_ref()
                .map(|value| value.binding_digest.as_str())
                != Some(expected_digest.as_str())
            || !valid_sha256(&receipt.receipt_digest)
        {
            return Err(Error::protocol(
                "ConsumeApproval returned an inconsistent receipt",
            ));
        }
        Ok(receipt)
    }
}

fn verify_binding(binding: &ApprovalBinding) -> Result<(), Error> {
    if !valid_sha256(&binding.intent_digest)
        || !valid_sha256(&binding.parameters_digest)
        || !valid_sha256(&binding.binding_digest)
    {
        return Err(Error::invalid_argument(
            "approval binding requires canonical SHA-256 digests",
        ));
    }
    let mut unsigned = binding.clone();
    let supplied = std::mem::take(&mut unsigned.binding_digest);
    let computed = format!("sha256:{:x}", Sha256::digest(unsigned.encode_to_vec()));
    if supplied != computed {
        return Err(Error::invalid_argument(
            "approval binding digest does not match its generated payload",
        ));
    }
    Ok(())
}

fn valid_timestamp(value: Option<&prost_types::Timestamp>) -> bool {
    value.is_some_and(|timestamp| {
        timestamp.seconds >= 0 && (0..1_000_000_000).contains(&timestamp.nanos)
    })
}
